WIP for financial support tool.

peter-fogg opened the pull request:
# [ECOM-2825](https://openedx.atlassian.net/browse/ECOM-2825)

Support tool for changing learner's enrollments. The code is ready to be reviewed, but I have to do some more styling work to get the tables all looking good. Given the size of this PR, I'd rather get reviews started now.

Sandbox will be at https://peter-fogg.sandbox.edx.org when it's finished building.

@bderusha @AlasdairSwan, can you review?

peter-fogg commented:
jenkins run bokchoy
jenkins run js
jenkins run lettuce

peter-fogg commented:
jenkins run js

AlasdairSwan commented on lms/djangoapps/support/static/support/js/enrollment_factory.js:
```diff
@@ -0,0 +1,13 @@
+        'support/js/views/enrollment'
+    ], function (_, EnrollmentView) {
+        return function (options) {
+            options = _.extend({el: $('.enrollment-content')}, options);
```
Why is ```el``` being set as a jQuery DOM object instead of a class? By default Backbone takes selectors and creates this.$el as a DOM object anyway.

peter-fogg replied:
Good point, fixed.

AlasdairSwan commented on lms/djangoapps/support/static/support/js/models/enrollment.js:
```diff
@@ -0,0 +1,24 @@
+                    success: _.bind(function (response) {
+                        this.set('manual_enrollment', response);
+                        this.set('mode', new_mode);
+                    }, this)
```
Error handling?

peter-fogg replied:
There's no error handling necessary on the model itself, but since it return the result of `$.ajax` error handlers can be attached by the calling code. This happens [in the modal view.](https://github.com/edx/edx-platform/pull/10966/files#diff-d8d718b7fbe13c7b74307820f77431c3R54)

AlasdairSwan commented on lms/djangoapps/support/static/support/js/views/enrollment.js:
```diff
@@ -0,0 +1,110 @@
+                        if (date) {
+                            return moment(date).format('MM/DD/YYYY (H:MM UTC)');
+                        }
+                        return 'N/A';
```
Nitpick: It may make the code clearer to have this as the else condition

peter-fogg replied:
Good idea, fixed.

AlasdairSwan commented on lms/djangoapps/support/static/support/js/views/enrollment.js:
```diff
@@ -0,0 +1,110 @@
+            search: function (event) {
+                event.preventDefault();
+                var username = this.getSearchString(),
+                    self = this;
```
var declarations should be at the very top of a function.

peter-fogg replied:
Fixed.

AlasdairSwan commented on lms/djangoapps/support/static/support/js/views/enrollment.js:
```diff
@@ -0,0 +1,110 @@
+                    success: function () {
+                        self.render();
+                    }
+                });
```
Shouldn't need to alias this, doesn't ```_.bind``` work here?

peter-fogg replied:
Fixed.

AlasdairSwan commented on lms/djangoapps/support/static/support/js/views/enrollment.js:
```diff
@@ -0,0 +1,110 @@
+                var button = $(event.currentTarget),
+                    course_id = button.data('course_id'),
+                    modes = button.data('modes').split(','),
+                    enrollment = this.enrollments.findWhere({course_id: course_id});
```
vars should be declared at the top of the function.

peter-fogg replied:
Fixed.

bderusha commented on lms/djangoapps/support/static/support/js/spec/collections/enrollment_spec.js:
```diff
@@ -0,0 +1,22 @@
+            });
+        });
+
+        it('sets its URL based on the user', function () {
```
:100: for not using "should" in the assertion

bderusha commented on lms/djangoapps/support/static/support/js/views/enrollment.js:
```diff
@@ -0,0 +1,110 @@
+    ], function (Backbone, _, moment, dataTables, EnrollmentModal, EnrollmentCollection, enrollmentTemplate) {
+        return Backbone.View.extend({
+
+            ENROLLMENT_CHANGE_REASONS: [
```
Should these also use `gettext`?

peter-fogg replied:
Hm, that's a good question. They need to be recorded un-translated, but it would make sense to display them translated.

bderusha commented on lms/djangoapps/support/tests/test_views.py:
```diff
@@ -1,13 +1,25 @@
 Tests for support views.
 """
 
+from datetime import datetime, timedelta
```
NIT: since `datetime` is part of std lib it should be on it's on line

peter-fogg replied:
Fixed (along with the other imports in this file).

peter-fogg commented:
For those of you following along at home, I've removed datatables in 3473da3 in order to let the tests pass. @bderusha @AlasdairSwan, I think I've addressed all your comments.

AlasdairSwan commented on lms/djangoapps/support/static/support/js/views/enrollment.js:
```diff
@@ -0,0 +1,103 @@
+                        if (!date) {
+                            return 'N/A';
+                        }
+                        return moment(date).format('MM/DD/YYYY (H:MM UTC)');
```
Nitpick: for clarity this could be changed to the else condition

AlasdairSwan commented on lms/djangoapps/support/static/support/js/views/enrollment_modal.js:
```diff
@@ -0,0 +1,76 @@
+            },
+
+            showErrors: function (errorMessage) {
+                this.$('.enrollment-change-errors').text(errorMessage).show();
```
Nitpick: jQuery's .hide and .show functions have a weird performance hit due to an edge case they cover. It would be cleaner to add/remove classes.

AlasdairSwan commented on lms/djangoapps/support/static/support/js/views/enrollment_modal.js:
```diff
@@ -0,0 +1,76 @@
+                    reason = this.$('.enrollment-reason').val() || this.$('.enrollment-reason-other').val();
+                event.preventDefault();
+                if (!reason) {
+                    this.showErrors(gettext("Please specify a reason."));
```
Nitpick: This should be single quotes for consistency within the file.

AlasdairSwan commented on lms/djangoapps/support/static/support/templates/enrollment-modal.underscore:
```diff
@@ -0,0 +1,27 @@
+<div class="enrollment-modal">
+  <h1 class="enrollment-change-title">Change enrollment for <%- enrollment.get('course_id') %></h1>
+  <p class="enrollment-change-errors" style="display: none;"></p>
+  <ol class="enrollment-info">
```
Why is the content being rendered as a list? Screenreaders are very verbose about entering lists which leads to poor user experience. I would recommend moving to divs.

AlasdairSwan commented on lms/djangoapps/support/static/support/templates/enrollment-modal.underscore:
```diff
@@ -0,0 +1,27 @@
+      </select>
+    </li>
+    <li class="enrollment-change-field">
+      <p><%- gettext("Reason for change:") %></p>
```
This should be a label for the the select

AlasdairSwan commented on lms/djangoapps/support/static/support/templates/enrollment-modal.underscore:
```diff
@@ -0,0 +1,27 @@
+          <option value="<%- reason %>"><%- translated_reason %></option>
+        <% }) %>
+      </select>
+      <input class="enrollment-reason-other" type="text" placeholder="<%- gettext('Explain if other.') %>" />
```
All form fields need labels.

AlasdairSwan commented on lms/djangoapps/support/static/support/templates/enrollment-modal.underscore:
```diff
@@ -0,0 +1,27 @@
+  <ol class="enrollment-info">
+    <li class="enrollment-change-field"><p><%- gettext("Current enrollment mode:") %> <%- enrollment.get('mode') %><p></li>
+    <li class="enrollment-change-field">
+      <p><%- gettext("New enrollment mode:") %></p>
```
This should be a label for the select

AlasdairSwan commented on lms/djangoapps/support/static/support/templates/enrollment-modal.underscore:
```diff
@@ -0,0 +1,27 @@
+  </ol>
+  <button class="enrollment-change-submit"><%- gettext("Submit enrollment change") %></button>
+  <button class="enrollment-change-cancel"><%- gettext("Cancel") %></button>
+</div>
```
This template is a form so I would be tempted to use a form element.

AlasdairSwan commented on lms/static/sass/views/_support.scss:
```diff
@@ -30,3 +30,117 @@
+
+    .enrollment-change-submit, .enrollment-change-cancel {
+      margin: ($baseline/4) auto;
+      @extend %t-action4;
```
extends should be declared at the top of the selector.

AlasdairSwan commented on lms/static/sass/views/_support.scss:
```diff
@@ -30,3 +30,117 @@
+
+  #enrollment-table_length, #enrollment-table_info {
+    display: inline-block;
+    @include float(left);
```
@includes should be added at the top of the selector.

AlasdairSwan commented on lms/static/sass/views/_support.scss:
```diff
@@ -30,3 +30,117 @@
+
+  #enrollment-table_filter, #enrollment-table_paginate {
+    display: inline-block;
+    @include float(right);
```
@includes should be added at the top of the selector.

AlasdairSwan commented on lms/static/sass/views/_support.scss:
```diff
@@ -30,3 +30,117 @@
+  .change-enrollment-btn, .change-enrollment-btn:hover {
+    margin: ($baseline/4) auto;
+    padding: ($baseline/4) 1px;
+    @extend %t-action4;
```
@extends and @includes should be added at the top of the selector.

AlasdairSwan commented:
Apologies @peter-fogg I should have caught the a11y HTML issues when I did my first round of reviews.

peter-fogg commented:
@AlasdairSwan I believe I've addressed all your feedback.

AlasdairSwan commented:
Nice :+1:

bderusha commented:
:+1: Looks good
